### PR TITLE
LH Compatible with java 21

### PR DESCRIPTION
**Pull request by sauljabin**

### CODE REVIEW

File: .github/workflows/tests.yml
Comment (coltmcnealy-lh) [inline .github/workflows/tests.yml:97]: Do we really need to run a matrix for the server? I think we can just support only the latest version of Java SDK, especially since we will be distributing it as a docker image in production.

For the `sdk-java`, a matrix is a very good idea.

Comment (mijailr) [reply .github/workflows/tests.yml:97]: I agree, server just the main version, sdk-java down to 17.

Comment (sauljabin) [reply .github/workflows/tests.yml:97]: The latest LTS is java 21, so far let's have a matrix with 17 and 21, and then let's remove 17

**Comment by sauljabin** [comment]
@coltmcnealy-lh @mijailr so far is working, but some tests are failing in 21
This class is failing: WfSpecModelTest

**Comment by sauljabin** [comment]
we can continue with this when we have deleted and migrated the e2e tests

File: .github/workflows/tests.yml
Comment (coltmcnealy-lh) [reply .github/workflows/tests.yml]: sdk-java should continue to support Java 11. Kafka supports all the way down to Java 8. Some of our users may be really outdated; I would hate to lose a deal because we had one less step in our pipeline

File: server/src/test/java/e2e/ChildThreadTest.java
Comment (coltmcnealy-lh) [inline server/src/test/java/e2e/ChildThreadTest.java:81]: I think I found a problem, or at least behavior that is different than what I expected. I had a NodeRun whose SubNodeRun completed properly but the variable mutations failed. The result is:
- :smiley: The ThreadRun is in the ERROR state
- :thinking: There is a Failure with type VAR_MUTATION_ERROR on the NodeRun rather than VAR_MUTATION_ERROR
- :thinking_face: The NodeRun is in the status COMPLETED

I thought that we decided that since an Edge is part of a Node, and the VariableMutation lives on the Edge, then technically the VariableMutation is part of the NodeRun itself. So therefore the NodeRun should be in the ERROR status...what do you think?
My tests show that this is not necessarily the case. Maybe I need to review that PR I made in March which fixes this 

(Link to original slack message, private to LH employees: https://littlehorseworkspace.slack.com/archives/C060A2XCX9D/p1725819836840409)


[11:27](https://littlehorseworkspace.slack.com/archives/C060A2XCX9D/p1725820034527379)
Also, another problem is that the failure is of the type VAR_SUB_ERROR not VAR_MUTATION_ERROR. I don't know if we actually use the status VAR_MUTATION_ERROR. Maybe we can get rid of it?

File: test-utils/src/main/java/io/littlehorse/test/WfRunVerifier.java
Comment (coltmcnealy-lh) [inline test-utils/src/main/java/io/littlehorse/test/WfRunVerifier.java:77]: This was very broken...when we tried to add the step it would accept it and then  silently not check it.

**Review by coltmcnealy-lh — APPROVED** [review]
LOL, technically since this PR was originally opened by @sauljabin I can approve it :sweat_smile: 

However, let's wait till @eduwercamacaro has a look before I approve it and go with it.

File: .github/workflows/tests.yml
Comment (sauljabin) [inline .github/workflows/tests.yml:87]: not needed if you hard coded the java version, example:` java-version: 21`

File: .github/workflows/tests.yml
Comment (sauljabin) [inline .github/workflows/tests.yml:108]: same here

File: .github/workflows/tests.yml
Comment (sauljabin) [inline .github/workflows/tests.yml:119]: nice

File: test-utils/src/main/java/io/littlehorse/test/internal/StandaloneTestBootstrapper.java
Comment (sauljabin) [inline test-utils/src/main/java/io/littlehorse/test/internal/StandaloneTestBootstrapper.java:36]: Interesting

File: server/src/test/java/e2e/ChildThreadTest.java
Comment (eduwercamacaro) [reply server/src/test/java/e2e/ChildThreadTest.java:81]: Hazim had a similar conversation a few months ago, in which he mentioned that the LH server handles mutation failures in a confusing way. Given the current design, I believe this is the best mechanism for handling variable mutation failures, as the node itself didn't fail. For instance, consider a wfspec that defines a TaskNode, defines several other variables, and then mutates those variables.

```
wf.execute('my-silly-validations');

wf.addVariable(...)
// More variables here

wf.mutate(...)
// More mutations here
```

Debugging will be quite confusing at first, clients may believe that the 'my-silly-validations' node has failed, which isn't the case. Therefore, in this scenario, it's preferable to have an edge failure within the node, indicating that the node should be completed. If we really want to change this behavior, I recommend splitting nodes and edges into separate thread-level objects.

File: server/src/test/java/e2e/RescueThreadRunTest.java
Comment (eduwercamacaro) [inline server/src/test/java/e2e/RescueThreadRunTest.java:89]: In my opinion, this is not semantically correct and goes against the objectives of this test framework, which is to provide a way to describe a set of verifiable individual constraints that enable clients (server devs)  describe how the workflow should look after applying some inputs (external events, input variables, user task forms, etc.). What I see in the `thenVerifyAllTaskRuns` method is a way to break that promise, because this allows to include a big step that executes a lot of assertions in it. My suggestion here is to include separate steps with fixed node positions to verify each task run.

Comment (coltmcnealy-lh) [reply server/src/test/java/e2e/RescueThreadRunTest.java:89]: Can we make this change in another PR?

File: server/src/test/java/e2e/ChildThreadTest.java
Comment (coltmcnealy-lh) [reply server/src/test/java/e2e/ChildThreadTest.java:81]: There is a difference between the `SubNodeRun` and the `NodeRun`. I think the fact that the `NodeRun` has a `Failure` on it and yet its status is `COMPLETED` is bad. Especially because the `Mutation` is part of the `Node`

File: server/src/test/java/e2e/RescueThreadRunTest.java
Comment (eduwercamacaro) [reply server/src/test/java/e2e/RescueThreadRunTest.java:89]: Sure, we can create a separate PR for this.

**Review by eduwercamacaro — APPROVED** [review]

**Pull request merged**
